paddle/optimizer/Tensor.h

```diff
@@ -0,0 +1,49 @@
+#ifndef PADDLE_OPTIMIZER_TENSOR_H_
+#define PADDLE_OPTIMIZER_TENSOR_H_
+/**
+ * @brief tensor used by optimizer
+ */
+
+#include <string.h>
+#include "paddle/utils/Common.h"
+#include "paddle/utils/Logging.h"
+
+namespace paddle {
+namespace optimizer {
+
+template <class T>
+class TensorT {
+public:
+  TensorT(size_t h, size_t w, T* data) : height_(h), width_(w), data_(data_) {}
+  TensorT(T* data, int size) : height_(1), width_(size), data_(data) {}
+  TensorT(const TensorT& t)
+      : TensorT(1, t.size(), 0, t.get_buffer(), false, false) {}
+  TensorT& operator=(const TensorT& t) {
+    this->width_ = t.size();
+    this->data_ = t.get_buffer();
+  }
+  T* get_buffer() { return this->data_; }
+  T& operator[](const size_t idx) {
+    CHECK(idx >= 0 && idx < this->width_) << "out of index range";
+    return data_[idx];
+  }
+  T& operator[](const size_t idx) const {
+    CHECK(idx >= 0 && idx < this->width_) << "out of index range";
+    return data_[idx];
+  }
+  // TODO: replace with tensorshape
+  size_t size() const { return this->width_; }
+
+protected:
+  size_t height_;
+  size_t width_;
+  T* data_;
+};
+
+// TODO(zhihong): design problem of dynamic datatype, need to fix it
+typedef TensorT<real> Tensor;
```

Comment:
let me explain this part more clear. 
In the last PR, I just implement Tensor with template class,  but it is the wrong way. Because template class will be lead to some specific types in **compile time**( explain the reason in a meeting will be better). 
In fact, tensor should represent some sort of consistent memory,  and support dynamic type **in runtime**
e.g. https://github.com/caffe2/caffe2/blob/master/caffe2/core/tensor.h#L643
just like an enhanced  `any` type in boost library, we can merge this PR, and enrich tensor library in the future. At this moment, we only support **compile period** datatype here.